lottery/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+    lottery_history: BTreeMap<u32, ActorId>,
+    lottery_id: u32,
+}
+
+impl Lottery {
+    fn lottery_state(&mut self) -> bool {
+        self.lottery_state.lottery_started
+            && (self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration)
+                > exec::block_timestamp()
+    }
+
+    fn start_lottery(&mut self, duration: u64) {
+        if msg::source() == self.lottery_owner && !self.lottery_state() {
+            self.lottery_state.lottery_started = true;
+            self.lottery_state.lottery_start_time = exec::block_timestamp();
+            self.lottery_state.lottery_duration = duration;
+            self.lottery_id += 1;
+        }
+    }
+
+    fn enter(&mut self) {
+        if self.lottery_state() && msg::value() > 0 {
+            if self.players_timestamp.get(&msg::source()) == None {
+                let player = Player {
+                    player_id: msg::source(),
+                    balance: msg::value(),
+                };
+
+                let player_index = self.players.len() as u32;
+                self.players.insert(player_index, player);
+                self.players_timestamp
+                    .insert(msg::source(), exec::block_timestamp());
+                msg::reply(Event::PlayerAdded(player_index), 0);
+            } else {
+                debug!("Player {:?} already added", msg::source());
+            }
+        }
+    }
+
+    fn leave_lottery(&mut self, index: u32) {
+        if self.lottery_state() {
+            if let Some(player) = self.players.get(&index) {
+                if player.player_id == msg::source() {
+                    msg::send_bytes(player.player_id, b"LeaveLottery", player.balance);
+                    self.players.remove(&index);
+                    self.players_timestamp.remove(&msg::source());
+                }
+            }
+        }
+    }
+
+    fn add_value(&mut self, index: u32) {
+        if self.lottery_state() {
+            self.players.entry(index).and_modify(|item| {
+                if item.player_id == msg::source() {
+                    item.balance = item.balance.saturating_add(msg::value());
+                }
+            });
+        }
+    }
+
+    fn get_balance(&mut self, index: u32) {
+        if self.lottery_state() {
+            if let Some(player) = self.players.get(&index) {
+                msg::reply(Event::Balance(player.balance), 0);
+            }
+        }
+    }
+
+    fn get_players(&mut self) {
+        if self.lottery_state() && !self.players.is_empty() {
+            msg::reply(Event::Players(self.players.clone()), 0);
+        }
+    }
+
+    fn get_random_number(&mut self) -> u32 {
+        let timestamp: u64 = exec::block_timestamp();
+        let code_hash: sp_core::H256 = blake2_256(&timestamp.to_be_bytes()).into();
+        let u_buf = code_hash.to_fixed_bytes();
+        let mut number: u32 = 0;
+
+        for u in u_buf {
+            number += u as u32;
+        }
+
+        number
+    }
```

Comment:
I'm afraid this function doesn't look good enough to be pseudo-random generator.

Reply:
In this case, the best option is to get a number outside the blockchain. But there is no such possibility at the moment.

Reply:
I think that using just cropped BLAKE2 is more random than the sum of bytes (as you calculate actually the mean of the pseudo-random value).